tests/xharness/AppRunner.cs

```diff
@@ -654,174 +442,9 @@ public async Task<int> RunAsync ()
-									MainLog.FullPath,
-									harness.XmlJargon);
-							}
-
-							break;
-						}
-					} catch (Exception e) {
-						harness.Log (2, "Failed to process crash report '{1}': {0}", e.Message, crashLog.Description);
-					}
-				}
-				if (!string.IsNullOrEmpty (crash_reason)) {
-					if (crash_reason == "per-process-limit") {
-						FailureMessage = "Killed due to using too much memory (per-process-limit).";
-					} else {
-						FailureMessage = $"Killed by the OS ({crash_reason})";
-					}
-					if (harness.InCI) {
-						resultParser.GenerateFailure (
-							Logs,
-							"crash",
-							AppInformation.AppName,
-							variation,
-							$"App Crash {AppInformation.AppName} {variation}",
-							$"App crashed: {FailureMessage}",
-							MainLog.FullPath,
-							harness.XmlJargon);
-					}
-				} else if (launch_failure) {
-					// same as with a crash
-					FailureMessage = $"Launch failure";
-					if (harness.InCI) {
-						resultParser.GenerateFailure (
-							Logs,
-							"launch",
-							AppInformation.AppName,
-							variation,
-							$"App Launch {AppInformation.AppName} {variation} on {deviceName}",
-							$"{FailureMessage} on {deviceName}",
-							MainLog.FullPath,
-							XmlResultJargon.NUnitV3);
-					}
-				} else if (!isSimulator && crashed && string.IsNullOrEmpty (crash_reason) && harness.InCI) {
-					// this happens more that what we would like on devices, the main reason most of the time is that we have had netwoking problems and the
-					// tcp connection could not be stablished. We are going to report it as an error since we have not parsed the logs, evne when the app might have
-					// not crashed. We need to check the main_log to see if we do have an tcp issue or not
-					var isTcp = false;
-					using (var reader = new StreamReader (MainLog.FullPath)) {
-						string line;
-						while ((line = reader.ReadLine ()) != null) {
-							if (line.Contains ("Couldn't establish a TCP connection with any of the hostnames")) {
-								isTcp = true;
-								break;
-							}
-						}
-					}
-
-					if (isTcp) {
-						resultParser.GenerateFailure (Logs,
-							"tcp-connection",
-							AppInformation.AppName,
-							variation,
-							$"TcpConnection on {deviceName}",
-							$"Device {deviceName} could not reach the host over tcp.",
-							MainLog.FullPath,
-							harness.XmlJargon);
-					}
-				} else if (timed_out && harness.InCI) {
-					resultParser.GenerateFailure (Logs,
-						"timeout",
-						AppInformation.AppName,
-						variation,
-						$"App Timeout {AppInformation.AppName} {variation} on bot {deviceName}",
-						$"{AppInformation.AppName} {variation} Test run timed out after {timeout.TotalMinutes} minute(s) on bot {deviceName}.",
-						MainLog.FullPath,
-						harness.XmlJargon);
-				}
-			}
-
-			return success.Value ? 0 : 1;
+			// check the final status, copy all the required data
+			(Result, FailureMessage) = await testResult.ParseResult ();
```

Comment:
Much simpler, hide all the complicated logic. We set mlaunch, execute and let the class parse the results.